tachyon/crypto/commitments/merkle_tree/field_merkle_tree/field_merkle_tree_mmcs.h

```diff
@@ -63,6 +70,106 @@ class FieldMerkleTreeMMCS final
     return true;
   }
 
+  const std::vector<math::RowMajorMatrix<PrimeField>>& DoGetMatrices() const {
+    return field_merkle_tree_.leaves();
+  }
+
+  [[nodiscard]] bool DoCreateOpeningProof(
+      size_t index, std::vector<std::vector<PrimeField>>* openings,
+      Proof* proof) const {
+    size_t max_rows = this->GetMaxRow();
+    size_t log_max_rows = base::bits::Log2Ceiling(max_rows);
+
+    // TODO(chokobole): Is it able to be parallelized?
+    *openings = base::Map(
+        field_merkle_tree_.leaves(),
+        [log_max_rows, index](const math::RowMajorMatrix<PrimeField>& matrix) {
+          size_t log_rows =
+              base::bits::Log2Ceiling(static_cast<size_t>(matrix.rows()));
+          size_t bits_reduced = log_max_rows - log_rows;
+          size_t reduced_index = index >> bits_reduced;
+          return base::CreateVector(matrix.cols(),
+                                    [reduced_index, &matrix](size_t col) {
+                                      return matrix(reduced_index, col);
+                                    });
+        });
+
+    *proof = base::CreateVector(log_max_rows, [this, index](size_t i) {
+      // NOTE(chokobole): :Let v be |index >> i|. if v is even, v ^ 1 is v + 1.
+      // Otherwise, v ^ 1 is v - 1.
+      return field_merkle_tree_.digest_layers()[i][(index >> i) ^ 1];
+    });
+
+    return true;
+  }
+
+  [[nodiscard]] bool DoVerifyOpeningProof(
+      const Commitment& commitment,
+      absl::Span<const math::Dimensions> dimensions_list, size_t index,
+      absl::Span<const std::vector<PrimeField>> opened_values,
+      const Proof& proof) const {
+    std::vector<math::Dimensions> sorted_dimensions_list(
+        dimensions_list.begin(), dimensions_list.end());
+
+    pdqsort(sorted_dimensions_list.begin(), sorted_dimensions_list.end(),
+            [](math::Dimensions a, math::Dimensions b) {
+              return a.height > b.height;
+            });
+    absl::Span<const math::Dimensions> remaining_dimensions_list =
+        absl::MakeConstSpan(sorted_dimensions_list);
+    absl::Span<const std::vector<PrimeField>> remaining_opened_values =
+        absl::MakeConstSpan(opened_values);
+
+    size_t next_layer =
+        absl::bit_ceil(remaining_dimensions_list.front().height);
+    size_t next_layer_size = 1;
+    for (size_t i = 1; i < remaining_dimensions_list.size(); ++i) {
+      if (next_layer == absl::bit_ceil(static_cast<size_t>(
+                            remaining_dimensions_list[i].height))) {
+        ++next_layer_size;
+      } else {
+        break;
+      }
+    }
```

Comment:
There are currently three places where the `next_layer_size` is calculated with same logic. How about extracting this code into a separate function to increase code reusability?